src/main/scala/edu/ie3/simona/model/grid/GridModel.scala

```diff
@@ -592,6 +634,133 @@ case object GridModel {
+    case TransformerControlGroup(measurements, _, vMax, vMin) =>
+      buildTransformerControlGroupModel(
+        measurementUnitInput,
+        measurements,
+        vMax,
+        vMin
+      )
+  }.toSet
+
+  /** Build a single control group model. Currently, only limit violation
+    * prevention logic is captured: The nodal regulation need is equal to the
+    * voltage change needed to comply with the given thresholds
+    *
+    * @param measurementUnitInput
+    *   Collection of all known [[MeasurementUnitInput]] s
+    * @param measurementConfigs
+    *   Collection of all uuids, denoting which of the [[MeasurementUnitInput]]
+    *   s does belong to this control group
+    * @param vMax
+    *   Upper permissible voltage magnitude
+    * @param vMin
+    *   Lower permissible voltage magnitude
+    * @return
+    *   A [[ControlGroupModel]]
+    */
+  private def buildTransformerControlGroupModel(
+      measurementUnitInput: java.util.Set[MeasurementUnitInput],
+      measurementConfigs: List[String],
+      vMax: Double,
+      vMin: Double
+  ): ControlGroupModel = {
+    val nodeUuids =
+      determineNodeUuids(measurementUnitInput, measurementConfigs)
+    buildTransformerControlModels(nodeUuids, vMax, vMin)
+  }
+
+  /** Determine the uuids of the nodes to control
+    *
+    * @param measurementUnitInput
+    *   Collection of all known [[MeasurementUnitInput]] s
+    * @param measurementConfigs
+    *   Collection of all uuids, denoting which of the [[MeasurementUnitInput]]
+    *   s does belong to this control group
+    * @return
+    *   A set of relevant nodal uuids
+    */
+  private def determineNodeUuids(
+      measurementUnitInput: java.util.Set[MeasurementUnitInput],
+      measurementConfigs: List[String]
+  ): Set[UUID] = Set.from(
+    measurementUnitInput.asScala
+      .filter(input =>
+        measurementConfigs.contains(input.getUuid.toString) && input.getVMag
+      )
+      .map(_.getNode.getUuid)
+  )
+
+  /** Build a single control group model. Currently, only limit violation
+    * prevention logic is captured: The nodal regulation need is equal to the
+    * voltage change needed to comply with the given thresholds
+    *
+    * @param nodeUuids
+    *   Collection of all relevant node uuids
+    * @param vMax
+    *   Upper permissible voltage magnitude
+    * @param vMin
+    *   Lower permissible voltage magnitude
+    * @return
+    *   A [[ControlGroupModel]]
+    */
+  private def buildTransformerControlModels(
+      nodeUuids: Set[UUID],
+      vMax: Double,
+      vMin: Double
+  ): ControlGroupModel = {
+    /* Determine the voltage regulation criterion for each of the available nodes */
+    val nodeUuidToRegulationCriterion = nodeUuids.map { uuid =>
+      uuid -> { (complexVoltage: Complex) =>
+        {
+          val vMag = complexVoltage.abs
+          if (vMag > vMax)
+            Some(vMax - vMag)
+          else if (vMag < vMin)
+            Some(vMin - vMag)
+          else
```

Comment:
I don't think so. This determines the need for regulation. Since the else case apply when the voltage is inbetween the limits there is no need to regulate it. Or did I got you wrong?

Reply:
Oh yes, I was too fast here